Segment Replication - Fix ShardLockObtained error during corruption cases

mch2 opened the pull request:
### Description
This change fixes a bug where shards could not be recreated locally after corruption. This occured because the store was not decref'd to 0 if the commit on close would fail with a corruption exception.

This change handles failures during the commit on close and marks the store corrupt if it occurs.
This change also handles failures while fetching local metadata by wrapping them as a corruption exception - without this shards would indefinitely retry and fail replication events.  Rethrowing the exception as a corruption exception ensures the shard is properly failed and re-recovered.  For this to happen a file would have to be wiped directly from the store, but would require a node reboot to correct it.

Added some ITs that will test these corruption cases - primary sending faulty data to the replica and a file being wiped underneath the shard.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8286
Resolves https://github.com/opensearch-project/OpenSearch/issues/9966

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] GitHub issue/PR created in [OpenSearch documentation repo](https://github.com/opensearch-project/documentation-website) for the required public documentation changes (#[Issue/PR number])

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

mch2 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27006/
> * **CommitID:** [2d9262e](https://github.com/opensearch-project/OpenSearch/commit/2d9262e18debf6f3c64816fa186e481f77f33c51)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/10139

mch2 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27018/
> * **CommitID:** [95fa8c7](https://github.com/opensearch-project/OpenSearch/commit/95fa8c7225e28281e915be4db16baad694da3bf2)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

test i added [SegmentReplicationIT.testWipeSegmentBetweenSyncs](https://build.ci.opensearch.org/job/gradle-check/27018/testReport/junit/org.opensearch.indices.replication/SegmentReplicationIT/testWipeSegmentBetweenSyncs/) is showing to be flaky from CI - removing it.  I've done these steps manually and can see from an IT standpoint the issue is resolved.  Leaving the other IT that is not flaky.

andrross commented on server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:
```diff
@@ -437,13 +437,38 @@ protected final void closeNoLock(String reason, CountDownLatch closedLatch) {
-                closedLatch.countDown();
+                try {
+                    store.decRef();
+                    logger.warn("engine closed [{}]", reason);
```
warn?

andrross commented on server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java:
```diff
@@ -437,13 +437,38 @@ protected final void closeNoLock(String reason, CountDownLatch closedLatch) {
+                        }
+                    }
+                }
+                try {
+                    IOUtils.close(readerManager);
+                } catch (Exception e) {
+                    logger.warn("Failed to close reader manager");
+                }
+                try {
+                    IOUtils.close(translogManager);
+                } catch (Exception e) {
+                    logger.warn("Failed to close translog");
+                }
```
You're swallowing the exceptions here without logging the details. Are you sure this is preferred over `IOUtils.close(readerManager, translogManager)` which will accumulate any exceptions and they'll get logged in the catch block?

mch2 replied:
oops good catch.  You're right will leave as it was, as long as we move store decref to finally block it solves the issue.

andrross commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1776,6 +1777,8 @@ public Store.MetadataSnapshot snapshotStoreMetadata() throws IOException {
         try (final GatedCloseable<SegmentInfos> snapshot = getSegmentInfosSnapshot()) {
             return store.getSegmentMetadataMap(snapshot.get());
+        } catch (IOException e) {
+            throw new OpenSearchCorruptionException("Error fetching local metadata");
```
Javadoc is now incorrect. Also this feels a bit like a hack to me.

mch2 replied:
It is, fixing it.  SegRep itself will only fail the shard if OpenSearchCorruptionException is the reason for failure.  We can just look for CorruptIndexException rather than re-wrapping.

mch2 replied:
also what this fixes is not likely but would prefer to be defensive. A file would have to be deleted underneath us before segrep runs.  If that happens we can get in a loop of failure at this call and retrying indefinitely.

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1784,88 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+                .put("index.refresh_interval", -1)
+                .build()
+        );
+        ensureYellow(INDEX_NAME);
+        final String replicaNode = internalCluster().startDataOnlyNode();
+        ensureGreen(INDEX_NAME);
```
Is this multistep process necessary? Why not just create two nodes then create the index?

mch2 replied:
Without this we don't know which node contains the primary/replica.  We could do it in one go but would have to then make some calls to know which is which.

andrross replied:
Oh I see. Makes sense.

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1784,88 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+        primaryTransportService.addSendBehavior(
+            internalCluster().getInstance(TransportService.class, replicaNode),
+            (connection, requestId, action, request, options) -> {
+                if (action.equals(SegmentReplicationTargetService.Actions.FILE_CHUNK) && failed.get() == false) {
+                    failed.compareAndSet(false, true);
```
Annoying nitpick:

```suggestion
                if (action.equals(SegmentReplicationTargetService.Actions.FILE_CHUNK) && failed.getAndSet(true) == false) {
```

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1784,88 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+        final IndexShard indexShard = getIndexShard(replicaNode, INDEX_NAME);
+        assertBusy(() -> {
+            // wait until the original shard is closed.
+            assertEquals(IndexShardState.CLOSED, indexShard.state());
+        }, 1, TimeUnit.MINUTES);
+        waitForActiveShardOnNode(replicaNode);
```
Is there a race here? How do you know you will observe that the shard is closed? A casual reading of this suggests that the replica shard could in theory recover before asserting that it is closed.

mch2 replied:
hmm yeah its not guaranteed and there is a race.  I will yank this for now and try and figure out a better way to test these types of scenarios.

mch2 replied:
I first tried a simple wait & assert on doc count - but queries would trigger while the shard is recovering and result in "all shards failed". oof

andrross replied:
You just want to assert that a recovery happens, right? I think you could call `GET <index>/_recovery` before the corruption and and then poll it after the corruption to verify that a new recovery did occur and that all documents are searchable afterwards. You don't need to observe the recovery happening in real time. There may be simpler mechanisms to assert that recovery happens as well.

mch2 replied:
aha this works! thanks.  I updated to query for stop time on a peer recovery and that a new recovery exists with a different time.

mch2 replied:
also reintroduced the ITs I removed...

mch2 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testWipeSegmentBetweenSyncs
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27078/
> * **CommitID:** [9dfc85d](https://github.com/opensearch-project/OpenSearch/commit/9dfc85d6452d300ace3dcc043bd4e0cfd9dfdc24)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

grr this IT is still showing flaky, yanking it.

mch2 commented:
> > ### Gradle Check (Jenkins) Run Completed with:
> > 
> > * **RESULT:** UNSTABLE ❕
> > * **TEST FAILURES:**
> > 
> > ```
> >       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testWipeSegmentBetweenSyncs
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * **URL:** https://build.ci.opensearch.org/job/gradle-check/27078/
> > * **CommitID:** [9dfc85d](https://github.com/opensearch-project/OpenSearch/commit/9dfc85d6452d300ace3dcc043bd4e0cfd9dfdc24)
> >   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.
> 
> grr this IT is still showing flaky, yanking it.

From this trace it looks like multiple nodes are spinning up node_t0-node_t4.  I haven't seen this locally but looks like the test is spawning additional data nodes instead of recovering on the original node/shard path.

mch2 commented:
> > > ### Gradle Check (Jenkins) Run Completed with:
> > > 
> > > * **RESULT:** UNSTABLE ❕
> > > * **TEST FAILURES:**
> > > 
> > > ```
> > >       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testWipeSegmentBetweenSyncs
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > * **URL:** https://build.ci.opensearch.org/job/gradle-check/27078/
> > > * **CommitID:** [9dfc85d](https://github.com/opensearch-project/OpenSearch/commit/9dfc85d6452d300ace3dcc043bd4e0cfd9dfdc24)
> > >   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.
> > 
> > 
> > grr this IT is still showing flaky, yanking it.
> 
> From this trace it looks like multiple nodes are spinning up node_t0-node_t4. I haven't seen this locally but looks like the test is spawning additional data nodes instead of recovering on the original node/shard path.

Ok found the cause here and fixing it in next commit - The issue appears to occur when a flush fails after we cause corruption - if during SR this will throw a FlushFailedException which is not treated as corruption case and assume the ReplicationTarget threads have already been cancelled.  This case needs to be treated as corruption so the shard is failed.

mch2 commented:
> > > > ### Gradle Check (Jenkins) Run Completed with:
> > > > 
> > > > * **RESULT:** UNSTABLE ❕
> > > > * **TEST FAILURES:**
> > > > 
> > > > ```
> > > >       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testWipeSegmentBetweenSyncs
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 
> > > > * **URL:** https://build.ci.opensearch.org/job/gradle-check/27078/
> > > > * **CommitID:** [9dfc85d](https://github.com/opensearch-project/OpenSearch/commit/9dfc85d6452d300ace3dcc043bd4e0cfd9dfdc24)
> > > >   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.
> > > 
> > > 
> > > grr this IT is still showing flaky, yanking it.
> > 
> > 
> > From this trace it looks like multiple nodes are spinning up node_t0-node_t4. I haven't seen this locally but looks like the test is spawning additional data nodes instead of recovering on the original node/shard path.
> 
> Ok found the cause here and fixing it in next commit - The issue appears to occur when a flush fails after we cause corruption - if during SR this will throw a FlushFailedException which is not treated as corruption case and assume the ReplicationTarget threads have already been cancelled. This case needs to be treated as corruption so the shard is failed.

Fixed this by ensuring we fail engines on corruption if a flush fails in NRTReplicationEngine.  Also cleaned up exception handling in finalizeReplication method.

mch2 commented on server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java:
```diff
@@ -279,14 +276,14 @@ private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse)
             // This can happen when the engine is closed in a separate thread.
             logger.warn("Shard is already closed, closing replication");
-        } catch (OpenSearchException ex) {
+        } catch (CancellableThreads.ExecutionCancelledException ex) {
```
note - I've cleaned up the handling here.  This path was incorrectly assuming any OSE was a cancellation and anything that was not an OSE was a corruption error.

It will now 1. Check for corruption and handle it by throwing a corruption exception.  2. Check for a cancellation and handle it (swallow the error so its not reported as a failure) or 3. throw it as a regular failure exception.  Anything not explicitly caught as corruption will be bubbled as a replication failed (non catastrophic/shard failure) exception.

mch2 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❕
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testMultipleShards
> ```
> 
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/27171/
> * **CommitID:** [2b80149](https://github.com/opensearch-project/OpenSearch/commit/2b80149af12743aa74d993f5d5849c44cffc0325)
>   Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

https://github.com/opensearch-project/OpenSearch/issues/9712

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1782,134 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+                }
+            }
+        );
+        for (int i = 0; i < 100; i++) {
+            client().prepareIndex(INDEX_NAME)
+                .setId(String.valueOf(i))
+                .setSource(jsonBuilder().startObject().field("field", i).endObject())
+                .get();
+        }
```
Nitpick, is it necessary to index 100 documents individually, blocking after each one? This could certainly be done faster.

mch2 replied:
We only need a single doc per refresh cycle for this test.  So yes it can be cut down.

That reminds me - we can likely cut down the execution times of SegmentReplicationUsingRemoteStore significantly but cutting down the docs indexed in this class's tests.  Those take much longer per doc given all are synced to the store.

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1782,134 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+                .get();
+        }
+        final long originalRecoveryTime = getRecoveryStopTime(replicaNode);
+        assertNotEquals(originalRecoveryTime, 0);
```
What case is this protecting against? Ideally we should not make any assumptions about the test environment's system clock, but in practice I'm sure this will never happen.

mch2 replied:
Was using this to ensure a new recovery existed for the node & it completed.  We could instead look for a recovery id for the specific node and check its state.  Probably easier to reason about.

andrross commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -1777,4 +1782,134 @@ public void testRealtimeTermVectorRequestsUnSuccessful() throws IOException {
+                .setSource(jsonBuilder().startObject().field("field", i).endObject())
+                .get();
+        }
+        final long originalRecoveryTime = getRecoveryStopTime(replicaNode);
```
I'm guessing the refresh interval of `-1` means that corruption can't happen until the refresh is triggered, but is there any reason you can't grab this "original recovery time" before you add the corruption behavior just to avoid the appearance of a possible race condition?

mch2 replied:
yes we can.  Will follow up with this.

andrross reviewed (approved):
Just some annoying test nitpicks that can be followed up (or ignored) in a subsequent PR
